Sources/NIOExtras/JSONRPCFraming+ContentLengthHeader.swift

```diff
@@ -0,0 +1,189 @@
+            case waitingForHeaderValue(name: String)
+            /// ... or for the payload of a given size.
+            case waitingForPayload(length: Int)
+        }
+
+        /// A `DecodingError` is sent through the pipeline if anything went wrong.
+        public enum DecodingError: Error, Equatable {
+            /// Missing 'Content-Length' header.
+            case missingContentLengthHeader
+
+            /// The value of the 'Content-Length' header was illegal, for example a negative number.
+            case illegalContentLengthHeaderValue(String)
+        }
+
+        public init() {}
+
+        // We start waiting for a header field (or the end of a header block).
+        private var state: State = .waitingForHeaderNameOrHeaderBlockEnd
+        private var payloadLength: Int? = nil
+
+        // `decode` will be invoked whenever there is more data available (or if we return `.continue`).
+        public mutating func decode(context: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
+            switch self.state {
+            case .waitingForHeaderNameOrHeaderBlockEnd:
+                // Given that we're waiting for the end of a header block or a new header field, it's sensible to
+                // check if this might be the end of the block.
+                if buffer.readableBytesView.starts(with: "\r\n".utf8) {
+                    buffer.moveReaderIndex(forwardBy: 2) // skip \r\n\r\n
+                    if let payloadLength = self.payloadLength {
+                        if payloadLength == 0 {
+                            // special case, we're not actually waiting for anything if it's 0 bytes...
+                            self.state = .waitingForHeaderNameOrHeaderBlockEnd
+                            self.payloadLength = nil
+                            context.fireChannelRead(self.wrapInboundOut(context.channel.allocator.buffer(capacity: 0)))
+                            return .continue
+                        }
+                        // cool, let's just shift to the `.waitingForPayload` state and continue.
+                        self.state = .waitingForPayload(length: payloadLength)
+                        self.payloadLength = nil
+                        return .continue
+                    } else {
+                        // this happens if we reached the end of the header block but we haven't seen the Content-Length
+                        // header, that's an error. It will be sent through the `Channel` and decoder won't be called
+                        // again.
+                        throw DecodingError.missingContentLengthHeader
+                    }
+                }
+
+                // Given that this is not the end of a header block, it must be a new header field. A new header field
+                // must always have a colon (or we don't have enough data).
+                if let colonIndex = buffer.readableBytesView.firstIndex(of: UInt8(ascii: ":")) {
```

Comment:
would caching `UInt8(ascii: ":")` be more efficient?

Reply:
@tomerd no, it would be less efficient. The compiler fully evaluates `UInt8(ascii: ":")` into whatever number it is.

Reply:
cool, til

Reply:
@tomerd just in case you want to try those things out yourself:

Step 1) Create a new swift file and put something like this in there:

```swift
@inline(never)
func SOMETHING_VERY_EASY_TO_GREP() -> Int {
    return Int(UInt8(ascii: "X"))
}

print(SOMETHING_VERY_EASY_TO_GREP())
```

then run this command:

```
$ swiftc -O -emit-assembly test.swift | grep -A10 SOMETHING_VERY_EASY
	.private_extern	_$s4test27SOMETHING_VERY_EASY_TO_GREPSiyF
	.globl	_$s4test27SOMETHING_VERY_EASY_TO_GREPSiyF
	.p2align	4, 0x90
_$s4test27SOMETHING_VERY_EASY_TO_GREPSiyF:
	pushq	%rbp
	movq	%rsp, %rbp
	movl	$88, %eax
	popq	%rbp
	retq

	.private_extern	_$ss23_ContiguousArrayStorageCyypGML
	.section	__DATA,__data
	.globl	_$ss23_ContiguousArrayStorageCyypGML
	.weak_definition	_$ss23_ContiguousArrayStorageCyypGML
```

and as you can see, the whole code of the function (modulo function start/end that would normally be inlined hadn't we done `@inline(never)` is

```
movl	$88, %eax
```

which just moves `88` into `eax`/`rax` which is the return register and 88 ascii is `X` 🙌

Reply:
Summary: The compiler fully converts `Int(UInt(ascii: "X"))` to the constant `88` which is puts in where ever you use it.

Reply:
👍 when feeling lazy you can also use (though less control about specific compiler version): https://godbolt.org

![Screen Shot 2019-05-14 at 11 13 17](https://user-images.githubusercontent.com/120979/57686140-59019480-7639-11e9-979a-08f86aa5113c.png)

It's fun for exploring what gets generated.

Reply:
@ktoso yes, good point! godbolt is amazing, I always forget to use it and use `grep`/Hopper :P